File: android/src/main/java/com/amplitude/skylab/Variant.java
package com.amplitude.skylab;

import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONException;
import org.json.JSONObject;

public class Variant {

    @Nullable public final String value;
    @Nullable public final Object payload;

    public Variant(@Nullable String value) {
        this(value, null);
    }

    public Variant(@Nullable String value, @Nullable Object payload) {
        this.value = value;
        this.payload = payload;
    }

    @Nullable
    @Deprecated
    public String value() {
        return this.value;
    }

    @Nullable
    @Deprecated
    public Object payload() {
        return this.payload;
    }

    @Override
    @NotNull
    public String toString() {
        return "Variant{" +
                "value='" + value + '\'' +
                ", payload=" + payload +
                '}';
    }

    @NotNull
    public static Variant fromJsonObject(@NotNull JSONObject variantJsonObj) throws JSONException {
        String value;

        if (variantJsonObj.has("value")) {
            value = variantJsonObj.getString("value");
        } else if (variantJsonObj.has("key")) {
            value = variantJsonObj.getString("key");
        } else {
            return new Variant(null);
        }

        Object payload = null;
        if (variantJsonObj.has("payload")) {
            payload = variantJsonObj.get("payload");
        }
        return new Variant(value, payload);
    }

    @NotNull
    public String toJson() {
        // create a JSONObject and then serialize it
        JSONObject jsonObj = new JSONObject();
        try {
            if (value != null) {
                jsonObj.put("value", value);
            }
            if (payload != null) {
                jsonObj.put("payload", payload);
            }
        } catch (JSONException e) {
            Log.w(Skylab.TAG, "Error converting Variant to json string", e);
        }

        return jsonObj.toString();
    }

    @NotNull
    public static Variant fromJson(@Nullable String json) {
        // deserialize into a JSONObject and then create a Variant
        if (json == null) {
            return new Variant(null, null);
        }

        try {
            JSONObject jsonObj = new JSONObject(json);
            return Variant.fromJsonObject(jsonObj);
        } catch (JSONException e) {
            // values persisted in older versions would throw a JSONException
            return new Variant(json);
        }
    }

    /**
     * See {@code #equals(Object)}
     */
    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((value == null) ? 0 : value.hashCode());
        return result;
    }

    /**
     * Variants are equal if their keys are equal
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Variant other = (Variant) obj;
        if (value == null) {
            return other.value == null;
        } else {
            return value.equals(other.value);
        }
    }
}

File: sdk/src/main/java/com/amplitude/experiment/util/Variant.kt
package com.amplitude.experiment.util

import com.amplitude.experiment.Variant
import com.amplitude.experiment.evaluation.EvaluationVariant
import org.json.JSONException
import org.json.JSONObject

internal fun Variant.toJson(): String {
    val jsonObject = JSONObject()
    try {
        jsonObject.put("key", key)
        if (value != null) {
            jsonObject.put("value", value)
        }
        if (payload != null) {
            jsonObject.put("payload", payload)
        }
        if (expKey != null) {
            jsonObject.put("expKey", expKey)
        }
        if (metadata != null) {
            jsonObject.put("metadata", metadata)
        }
    } catch (e: JSONException) {
        Logger.w("Error converting Variant to json string", e)
    }
    return jsonObject.toString()
}

internal fun String?.toVariant(): Variant? {
    return if (this == null) {
        null
    } else {
        JSONObject(this).toVariant()
    }
}

internal fun JSONObject?.toVariant(): Variant? {
    return if (this == null) {
        null
    } else try {
        var key = when {
            has("key") -> getString("key")
            else -> null
        }
        val value = when {
            has("value") -> getString("value")
            else -> null
        }

        if (key == null && value == null) {
            return null
        }
        if (key == null && value != null) {
            key = value
        }

        val payload = when {
            has("payload") -> {
                val payload = get("payload")
                if (payload is JSONObject) {
                    payload.toMap()
                } else {
                    payload
                }
            }

            else -> null
        }

        var expKey = when {
            has("expKey") -> getString("expKey")
            else -> null
        }
        var metadata = when {
            has("metadata") -> getJSONObject("metadata").toMap()
            else -> null
        }?.toMutableMap()
        if (metadata != null && metadata["experimentKey"] != null) {
            expKey = metadata["experimentKey"] as? String
        } else if (expKey != null) {
            metadata = metadata ?: HashMap()
            metadata["experimentKey"] = expKey
        }

        Variant(value, payload, expKey, key, metadata)
    } catch (e: JSONException) {
        Logger.w("Error parsing Variant from json string $this")
        null
    }
}

internal fun EvaluationVariant.convertToVariant() : Variant {
    val experimentKey = this.metadata?.get("experimentKey")?.toString()
    val value = when {
        this.value != null -> this.value.toString()
        else -> null
    }
    val expKey = when {
        experimentKey != null -> experimentKey
        else -> null
    }
    val payload = when {
        this.payload != null -> this.payload
        else -> null
    }
    val metadata = when {
        this.metadata != null -> this.metadata
        else -> null
    }
    return Variant(value, payload, expKey, this.key, metadata)
}
